### ISSUE TITLE
MUC Self-Ping: stack-trace / no return after outage

### ISSUE THREAD

**Issue by ge0rg** [horazont/aioxmpp#312]
The current code is exhibiting a stacktrace and not coming back to a MUC after an outage. It also looks like it tries to leave the "broken" room twice:

```
2019-09-23 01:09:48,232 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: soft-limit tripped, starting pinger
2019-09-23 01:09:48,233 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: request to start pinger
2019-09-23 01:09:48,234 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: request to stop pinger
2019-09-23 01:09:48,235 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: already stopped
2019-09-23 01:09:48,236 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger booted up
2019-09-23 01:09:48,237 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger loop. interval=datetime.timedelta(0, 120)
2019-09-23 01:09:48,238 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: sending self-ping with timeoutdatetime.timedelta(0, 480)
2019-09-23 01:09:48,239 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger waiting for 1 pings forat most 120.0s
2019-09-23 01:09:48,240 DEBUG aioxmpp.StanzaStream sending <iq from=None to='room@chat.yax.im/roombot' id=':f-3N6imBswtvm7Izcvg3' type=<IQType.GET: 'get'> data=<aioxmpp.ping.xso.Ping object at 0xb28f7c8c>> and waiting for it to be sent
2019-09-23 01:09:48,241 DEBUG aioxmpp.StanzaStream enqueued stanza <iq from=None to='room@chat.yax.im/roombot' id=':f-3N6imBswtvm7Izcvg3' type=<IQType.GET: 'get'> data=<aioxmpp.ping.xso.Ping object at 0xb28f7c8c>> with token <StanzaToken id=0x00000000b28f52fc>
2019-09-23 01:09:48,243 DEBUG aioxmpp.StanzaStream forwarding stanza to xmlstream: <iq from=None to='room@chat.yax.im/roombot' id=':f-3N6imBswtvm7Izcvg3' type=<IQType.GET: 'get'> data=<aioxmpp.ping.xso.Ping object at 0xb28f7c8c>>
2019-09-23 01:09:48,245 DEBUG aioxmpp.node.Client.XMLStream SENT b'<iq type="get" id=":f-3N6imBswtvm7Izcvg3" to="room@chat.yax.im/roombot"><ping xmlns="urn:xmpp:ping"/></iq>'
2019-09-23 01:09:48,246 DEBUG aioxmpp.StanzaStream sending SM req
2019-09-23 01:09:48,247 DEBUG aioxmpp.node.Client.XMLStream SENT b'<r xmlns="urn:xmpp:sm:3"/>'
2019-09-23 01:09:48,249 DEBUG aioopenssl.trace.fd=7 _write_ready: nothing more to write, removing writer
2019-09-23 01:09:48,254 DEBUG aioxmpp.node.Client.XMLStream RECV b"<a h='735' xmlns='urn:xmpp:sm:3'/>"
2019-09-23 01:09:48,256 DEBUG aioxmpp.StanzaStream received SM ack: <aioxmpp.nonza.SMAcknowledgement counter=735 at 0xb28f7e2c>
2019-09-23 01:09:48,256 DEBUG aioxmpp.StanzaStream sm_ack(735)
2019-09-23 01:09:48,257 DEBUG aioxmpp.StanzaStream sm_ack: to drop 1, unacked: 1
2019-09-23 01:09:48,257 DEBUG aioxmpp.StanzaStream 1 stanzas acked by remote
2019-09-23 01:09:48,291 DEBUG aioxmpp.node.Client.XMLStream RECV b"<iq to='bot@domain/hxUBuMjX' id=':f-3N6imBswtvm7Izcvg3' from='room@chat.yax.im/roombot' type='error'><error type='cancel' by='domain'><remote-server-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/><text xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Server-to-server connection failed: Connecting failed: connection refused</text></error></iq><r xmlns='urn:xmpp:sm:3'/>"
2019-09-23 01:09:48,369 DEBUG aioxmpp.StanzaStream incoming iq: <iq from='room@chat.yax.im/roombot' to='bot@domain/hxUBuMjX' id=':f-3N6imBswtvm7Izcvg3' type=<IQType.ERROR: 'error'> error=<remote-server-not-found type=<ErrorType.CANCEL: 'cancel'> text='Server-to-server connection failed: Connecting failed: connection refused'>>
2019-09-23 01:09:48,377 DEBUG aioxmpp.StanzaStream iq is response
2019-09-23 01:09:48,693 DEBUG aioxmpp.StanzaStream iq response delivered to key (JID(localpart='room', domain='chat.yax.im', resource='roombot'), ':f-3N6imBswtvm7Izcvg3')
2019-09-23 01:09:48,703 DEBUG aioxmpp.StanzaStream received SM request: <aioxmpp.nonza.SMRequest object at 0xb28f78ec>
2019-09-23 01:09:48,704 DEBUG aioxmpp.StanzaStream sending SM ack: <aioxmpp.nonza.SMAcknowledgement counter=4741 at 0xb3a52b2c>
2019-09-23 01:09:48,705 DEBUG aioxmpp.node.Client.XMLStream SENT b'<a xmlns="urn:xmpp:sm:3" h="4741"/>'
2019-09-23 01:09:48,707 DEBUG aioopenssl.trace.fd=7 _write_ready: nothing more to write, removing writer
2019-09-23 01:09:48,710 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: ping reply has error indicating that the client got removed: ErrorCondition.REMOTE_SERVER_NOT_FOUND
2019-09-23 01:09:48,711 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: exited detected
2019-09-23 01:09:48,711 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: request to stop pinger
2019-09-23 01:09:48,712 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: sending cancel signal
2019-09-23 01:09:48,783 DEBUG aioxmpp.StanzaStream enqueued stanza <presence from=None to='room@chat.yax.im/roombot' id=':izv3ytfGrfEcRNhUCkmW' type=<PresenceType.UNAVAILABLE: 'unavailable'>> with token <StanzaToken id=0x00000000b28aedc4>
2019-09-23 01:09:48,785 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: disabled monitoring
2019-09-23 01:09:48,785 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: request to stop pinger
2019-09-23 01:09:48,786 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: already stopped
2019-09-23 01:09:48,831 DEBUG aioxmpp.StanzaStream enqueued stanza <presence from=None to='room@chat.yax.im/roombot' id=':ncvnFr2CPND9O5c_pIyW' type=<PresenceType.AVAILABLE: None>> with token <StanzaToken id=0x00000000b28ad66c>
2019-09-23 01:09:48,832 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger loop. interval=datetime.timedelta(0, 120)
2019-09-23 01:09:48,832 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger has nothing to do, sleeping for 123.5838942527771
2019-09-23 01:09:48,833 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor room@chat.yax.im/roombot: pinger exited
Traceback (most recent call last):
  File "/usr/src/aioxmpp/aioxmpp/muc/self_ping.py", line 269, in _pinger
    yield from asyncio.sleep(timeout)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 524, in sleep
    return (yield from future)
  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 304, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 285, in result
    raise CancelledError
concurrent.futures._base.CancelledError
2019-09-23 01:09:48,843 DEBUG aioxmpp.StanzaStream forwarding stanza to xmlstream: <presence from=None to='room@chat.yax.im/roombot' id=':izv3ytfGrfEcRNhUCkmW' type=<PresenceType.UNAVAILABLE: 'unavailable'>>
2019-09-23 01:09:49,008 DEBUG aioxmpp.node.Client.XMLStream SENT b'<presence type="unavailable" id=":izv3ytfGrfEcRNhUCkmW" to="room@chat.yax.im/roombot"><x xmlns="http://jabber.org/protocol/muc"/></presence>'
2019-09-23 01:09:49,008 DEBUG aioxmpp.StanzaStream forwarding stanza to xmlstream: <presence from=None to='room@chat.yax.im/roombot' id=':ncvnFr2CPND9O5c_pIyW' type=<PresenceType.AVAILABLE: None>>
2019-09-23 01:09:49,013 DEBUG aioxmpp.node.Client.XMLStream SENT b'<presence id=":ncvnFr2CPND9O5c_pIyW" to="room@chat.yax.im/roombot"><x xmlns="http://jabber.org/protocol/muc"><history maxchars="0" maxstanzas="0"/></x></presence>'
2019-09-23 01:09:49,013 DEBUG aioxmpp.StanzaStream sending SM req
2019-09-23 01:09:49,020 DEBUG aioxmpp.node.Client.XMLStream SENT b'<r xmlns="urn:xmpp:sm:3"/>'
2019-09-23 01:09:49,022 DEBUG aioopenssl.trace.fd=7 _write_ready: nothing more to write, removing writer
2019-09-23 01:09:49,032 DEBUG aioxmpp.node.Client.XMLStream RECV b"<a h='737' xmlns='urn:xmpp:sm:3'/>"
2019-09-23 01:09:49,034 DEBUG aioxmpp.StanzaStream received SM ack: <aioxmpp.nonza.SMAcknowledgement counter=737 at 0xb28f728c>
2019-09-23 01:09:49,035 DEBUG aioxmpp.StanzaStream sm_ack(737)
2019-09-23 01:09:49,036 DEBUG aioxmpp.StanzaStream sm_ack: to drop 2, unacked: 2
2019-09-23 01:09:49,037 DEBUG aioxmpp.StanzaStream 2 stanzas acked by remote
2019-09-23 01:09:49,045 DEBUG aioxmpp.node.Client.XMLStream RECV b"<presence to='bot@domain/hxUBuMjX' id=':izv3ytfGrfEcRNhUCkmW' from='room@chat.yax.im/roombot' type='error'><error type='cancel' by='domain'><remote-server-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/><text xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Server-to-server connection failed: Connecting failed: connection refused</text></error></presence><presence to='bot@domain/hxUBuMjX' id=':ncvnFr2CPND9O5c_pIyW' from='room@chat.yax.im/roombot' type='error'><error type='cancel' by='domain'><remote-server-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/><text xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Server-to-server connection failed: Connecting failed: connection refused</text></error></presence><r xmlns='urn:xmpp:sm:3'/>"
2019-09-23 01:09:49,059 DEBUG aioxmpp.StanzaStream incoming presence: <presence from='room@chat.yax.im/roombot' to='bot@domain/hxUBuMjX' id=':izv3ytfGrfEcRNhUCkmW' type=<PresenceType.ERROR: 'error'>>
2019-09-23 01:09:49,061 DEBUG aioxmpp.StanzaStream incoming presence: <presence from='room@chat.yax.im/roombot' to='bot@domain/hxUBuMjX' id=':ncvnFr2CPND9O5c_pIyW' type=<PresenceType.ERROR: 'error'>>
2019-09-23 01:09:49,064 DEBUG aioxmpp.StanzaStream received SM request: <aioxmpp.nonza.SMRequest object at 0xb3a6756c>
2019-09-23 01:09:49,065 DEBUG aioxmpp.StanzaStream sending SM ack: <aioxmpp.nonza.SMAcknowledgement counter=4743 at 0xb3a52b2c>
2019-09-23 01:09:49,067 DEBUG aioxmpp.node.Client.XMLStream SENT b'<a xmlns="urn:xmpp:sm:3" h="4743"/>'
2019-09-23 01:09:49,069 DEBUG aioopenssl.trace.fd=7 _write_ready: nothing more to write, removing writer
```

**Comment by horazont** [comment]
The crux is this one (added linebreaks for readability):

```
2019-09-23 01:09:48,710 DEBUG aioxmpp.node.Client.muc.MUCClient.MUCMonitor 
  room@chat.yax.im/roombot: ping reply has error indicating that the client got 
  removed: ErrorCondition.REMOTE_SERVER_NOT_FOUND
```

As per XEP-0410, any unexpected error should be treated as "client got removed". I made a PR to fix this: xsf/xeps#834.

aioxmpp notes that the room has been left, and then attempts to re-join (once!). However, the re-join fails because of remote-server-not-found and is not re-tried automatically.
